### PR TITLE
context parallelism

### PR DESCRIPTION
**Pull request by xrennvidia**
# What does this PR do ?

GPT training with long-context input (e.g., sequence length of 16, 32K, 64K) can easily overflow GPU memory with huge activations. Context parallelism splits long-context input along the dimension of sequence length, parallelizes partitioned sequence segments among multiple GPUs. In this way, each GPU only needs to store activations of a part of sequence length, so that we can avoid memory overflow.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Comment by xrennvidia** [comment]
jenkins

**Comment by xrennvidia** [comment]
jenkins

**Comment by stu1130** [comment]
@xrennvidia Hey Xiaowei, we are testing new CP feature with this PR. I saw your last commit "recover seq-length which has been fixed in mcore". Would you mind to tell us which commit of Megatron-LM && TransformerEngine I should use to have the fix? Thanks a lot!

**Comment by xrennvidia** [comment]
@stu1130 oh, sorry. You can keep the NeMo seq-length fix in your test now. I just moved this fix to mcore in my mcore PR, which has not been merged to MLM main branch yet. Basically in mcore [I divide tensor_shape[0] by CP size at here](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/pipeline_parallel/schedules.py#L460).

**Comment by stu1130** [comment]
Thanks a lot @xrennvidia Do you know when the PR will get merged into the master?

**Comment by xrennvidia** [comment]
@stu1130 we are doing internal tests now. We will merge it asap.

File: nemo/collections/nlp/models/language_modeling/megatron_base_model.py [hunk 1]
```diff
@@ -231,6 +232,27 @@ def setup_transformer_engine_tp_groups(self):
                     tp_group = parallel_state.get_tensor_model_parallel_group()
                     child.set_tensor_parallel_group(tp_group)
 
+    def setup_transformer_engine_cp_groups(self):
+        """ This should be called after context parallel groups have been initialized
+            and only needs to be called when using Transformer Engine.
+        """
+        cp_stream = torch.cuda.Stream()
+
+        for module in self.get_gpt_module_list():
```
Comment (stu1130) [inline nemo/collections/nlp/models/language_modeling/megatron_base_model.py:241]: couldn't find this method. Is it get_model_module_list?

Comment (xrennvidia) [reply nemo/collections/nlp/models/language_modeling/megatron_base_model.py:241]: please pull the latest commit. this stale code.

Comment (stu1130) [reply nemo/collections/nlp/models/language_modeling/megatron_base_model.py:241]: Thanks. Did you mean latest NeMo context parallel commit or Megatron-LM/TransformerEngine?

Comment (xrennvidia) [reply nemo/collections/nlp/models/language_modeling/megatron_base_model.py:241]: latest NeMo context parallel commit

Comment (stu1130) [reply nemo/collections/nlp/models/language_modeling/megatron_base_model.py]: Thanks!

**Comment by xrennvidia** [comment]
jenkins

**Comment by xrennvidia** [comment]
jenkins

**Comment by stu1130** [comment]
Hey @xrennvidia I am using Megatron-LM (mcore r0.4.0). If I pull the latest change in the PR, would I also need to cherry-pick https://github.com/NVIDIA/Megatron-LM/commit/5eaa937e562ee64775a6084e27e920f557e5709e to r0.4.0? Again, thanks for developing the feature, it really benefits our use case a lot!

**Comment by xrennvidia** [comment]
> Hey @xrennvidia I am using Megatron-LM (mcore r0.4.0). If I pull the latest change in the PR, would I also need to cherry-pick [NVIDIA/Megatron-LM@5eaa937](https://github.com/NVIDIA/Megatron-LM/commit/5eaa937e562ee64775a6084e27e920f557e5709e) to r0.4.0? Again, thanks for developing the feature, it really benefits our use case a lot!

Very happy to know this is helpful :) . If you want to run with PP > 1, you need to cherry-pick it. 

Also FYI, [I have a fix for your issue at here](https://github.com/NVIDIA/NeMo/pull/7739#issuecomment-1848008183). I think the fix should be merged to MLM main branch soon, maybe tomorrow.

**Comment by xrennvidia** [comment]
jenkins

**Review by ericharper — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**